### DISCUSSION TITLE
Use $XDG_STATE_HOME for temporary state data

### DISCUSSION THREAD

**Discussion by 62832** [jarun/nnn#1786] [Ideas]
The latest version (0.8, 08/05/2021) of the [XDG Base Directory specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html) introduces a new environment variable `$XDG_STATE_HOME` for "state" data that should persist between restarts but isn't necessary to backup or store in other locations such as "config" and "data". I feel like it might be worth making use of this for things like session files, `.selection` and `$NNN_TMPFILE` instead of the config directory.

**Comment by azuline** [comment]
I do not think this is worth a breaking change or parallel code path

**Reply by 62832** [reply]
I won't argue with whether it would be a breaking change and whether it would be worth it, but the way I see it, the spec isn't being followed properly having everything jammed into the config directory. At the very least, I've noticed now that `$NNN_SEL` exists to override the location of the `.selection` file, but it would still be nice to default to the state directory and use it for things like sessions as well.

**Reply by N-R-K** [reply]
> whether it would be a breaking change

It definitely would be a breaking change since lots of scripts/plugins have been written with the current path logic expectations.

> the spec isn't being followed properly

XDG spec is more of a convention than a spec that *must be* rigorously followed. In fact, most software I come across does not hard follow and they usually only supports DATA, CONFIG and CACHE dir or less.

Granted than nnn is following even less than most, only supporting CONFIG. But that's fine, we don't advertise full xdg compliance. And I also don't think it's worthwhile to aim for it.

**Reply by 62832** [reply]
My objection to this is the fact that I try to use the config directory as a relatively small directory to be able to back up remotely and retrieve on any other machine I please. If only one of the directories in `$HOME/.config/nnn` is *actually* for configuration of my own (presumably `plugins`), then that means there's needless clutter in the directory meant purely for configuration.

I can understand however that this isn't going to be prioritised any time soon by any of the existing maintainers, so would you at least be willing to accept any PRs that aim to add proper support?

**Comment by N-R-K** [comment]
> session files

You should be able to turn the session directory into a symlink so that the actual files are stored elsewhere.

> .selection and $NNN_TMPFILE

Both of these can be set to a custom path. (`$NNN_SEL` for selection file).

> so would you at least be willing to accept any PRs that aim to add proper support?

IMO no, since it'd be a breaking change.

- - -

For reference, this is what I have in my shell profile to take non-config files out of `~/.config`:

```sh
export NNN_TMPFILE="${XDG_CACHE_HOME}/nnn/lastdir" # cd on quit
export NNN_FIFO="${XDG_CACHE_HOME}/nnn/nnn.fifo" # temp file thingy
export NNN_SEL="${XDG_CACHE_HOME}/nnn/selection" # selection file
export NNN_FINDHIST="${XDG_CACHE_HOME}/nnn/find_hist" # for finder plugin
```

The `plugin` and `bookmarks` directories are symlinks that points elsewhere. I don't use sessions, but I suspect it should be possible to make it a symlink as well.
